src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -1420,12 +1421,31 @@ fn calculate_normal(
     }
     .to_vec();
 
+    // Include all the args from `[lints.rust.unexpected_cfgs.check-cfg]`
+    let mut lint_check_cfg = Vec::new();
+    if let Ok(Some(lints)) = unit.pkg.manifest().resolved_toml().resolved_lints() {
+        if let Some(rust_lints) = lints.get("rust") {
+            if let Some(unexpected_cfgs) = rust_lints.get("unexpected_cfgs") {
+                if let Some(config) = unexpected_cfgs.config() {
+                    if let Some(check_cfg) = config.get("check-cfg") {
+                        if let Ok(check_cfgs) =
+                            toml::Value::try_into::<Vec<String>>(check_cfg.clone())
+                        {
+                            lint_check_cfg = check_cfgs;
+                        }
+                    }
+                }
+            }
+        }
+    }
```

Comment:
Rather than re-looking up this entry, we should see if there is a good way to make this reuseable

Example areas to explore
- I assume we fingerprint the build script check-cfg.  Can we mix the two
- We already check `lint_rustflags`.  Should we instead move the lookup to be a part of that?

Reply:
> I assume we fingerprint the build script check-cfg. Can we mix the two

I don't know, I'm not sure we do anything special for build-script. I think they are just seen as deps in fingerprint.

> We already check lint_rustflags. Should we instead move the lookup to be a part of that?

`lint_rustflags` creates the different `--warn`, `--allow`, ... flags, which we could add the `--check-cfg` args at some point, but not right now since we need to gate all `--check-cfg` args behind the compiler supporting check-cfg stably, which can only be done when we have access to the `Unit` (I think) which is way after `lint_rustflags` is called

Reply:
Created #13975 (so the hack can be tracked) and updated the code with your suggested wording (as well as changing it to a `HACK`).